### DISCUSSION TITLE
With Slugs enabled Woocommerce dont show 404 page for products

### DISCUSSION THREAD

**Discussion by andreysneg** [qtranslate/qtranslate-xt#1250] [Q&A] [answered]
If Slugs enabled ( new extension or old slugs plugin ) Woocommerce does not show 404 page for missing products.

If product missing site go to Shop page.

But must show 404 page if product missing.  Its bad for SEO. 

How remove redirect to shop page ?

**Comment by andreysneg** [comment]
I found :  If Slug extension dont find product hi make global wp->is_archive  is true .  When wp dont show 404 page.

You can debug it in  class-wp.php  line 736 and found is_post_type_archive()= true.

```
if ( is_author() && is_numeric( $author ) && $author > 0 && is_user_member_of_blog( $author )
				// Don't 404 for these queries if they matched an object.
				|| ( is_tag() || is_category() || is_tax() || is_post_type_archive() ) && get_queried_object()
				// Don't 404 for these queries either.
				|| is_home() || is_search() || is_feed()
			) {
				$set_404 = false;
			}
```

Why  is_archive  is true ?  this is product.

**Reply by herrvigg** [reply]
Maybe @spleen1981 can help here if he finds some time?

**Comment by andreysneg** [comment]
File  modules/slugs/includes/class-qtranslate-slug.php  -   https://bit.ly/3THqk0C

module create QUERY and wordpress find archive here  file class-wp-query.php  -  https://bit.ly/3TTptdD

Becouse $post_type_obj has   :    has_archive="shop"  ,  wp think what is archive page and dont show 404 page -  https://bit.ly/3EyHwkA

What write here: https://bit.ly/3THqk0C
for show page 404 ?

I try, but ugly result -
   global $wp_query;
   $wp_query->set_404();
   status_header( 404 );
   get_template_part( 404 );
   exit();

**Comment by andreysneg** [comment]
There is a solution but not a beautiful one. Please make normal solution for this.

1. in file  class-qtranslate-slug.php   https://bit.ly/3GqNOo1


2. in own module or function.php insert code

```
add_action( 'wp', 'force_404' );
function force_404() {
    if (isset($_ENV['p404'] )){
        global $wp_query;
        $wp_query->set_404();
        status_header( 404 );
        nocache_headers();
    }
}
```

**Comment by spleen1981** [comment, marked as answer]
Moved to issues (#1260) , where it belongs.

**Reply by herrvigg** [reply]
Yes all of this should be is an issue, not a discussion. I close this one.

**Reply by herrvigg** [reply]
OK you alread did :)
